walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
+    boolean considerMandatoryAttendees = request.getAttendees().size() > 0;
+    boolean considerOptionalAttendees = request.getOptionalAttendees().size() > 0;
+
+    for (Event event : sortedEvents) {
+      TimeRange when = event.getWhen();
+      // If true, the meeting request cannot overlap with this.
+      boolean containsMandatoryAttendees = anyAttendeesAreMandatory(
+        event,
+        request
+      );
+      // If true, meeting requests that try to include optional attendees cannot overlap with this.
+      boolean containsOptionalAttendees = anyAttendeesAreOptional(
+        event,
+        request
+      );
+
+      if (containsMandatoryAttendees) {
+        hasMandatoryEvents = true;
+      }
+      if (containsOptionalAttendees) {
+        hasOptionalEvents = true;
+      }
+
+      // As mentioned above, we track 'mandatory' people and 'optional' people as two classes.
+      // For a given class, a window exists if there is time between current event involving that class, 
+      // And the end of the last event involving that class (_____BlockExtent)
+      boolean optionalWindowExists = when.start() > optionalBlockExtent;
+      boolean mandatoryWindowExists = when.start() > mandatoryBlockExtent;
+
+      // Makes sure the requested meeting length fits within the window.
+      boolean optionalWindowLongEnough = optionalWindowExists && windowIsLongEnough(optionalBlockExtent, when.start(), request);
+      boolean mandatoryWindowLongEnough = mandatoryWindowExists && windowIsLongEnough(mandatoryBlockExtent, when.start(), request);
+
+      // Optional-based windows must also include all mandatory attendees.
+      if (considerOptionalAttendees && optionalWindowLongEnough && mandatoryWindowLongEnough) {
+        timeSlotsWithOptional.add(
+          /* Add non-inclusive window because it can't overlap with the start of this event. */
+          TimeRange.fromStartEnd(
+            Math.max(optionalBlockExtent, mandatoryBlockExtent), // Only use overlap between optional/mandatory availability.
+            when.start(), false)
+        );
+      }
+      
+
+      if (containsMandatoryAttendees && mandatoryWindowLongEnough) {
+        timeSlots.add(
+          /* Add non-inclusive window because it can't overlap with the start of this event. */
+          TimeRange.fromStartEnd(mandatoryBlockExtent, when.start(), false)
+        );
+      }
+
+      // If this window contains members from a certain class, update their respective block extent.
+      if (containsOptionalAttendees) { // We group all optional attendees as one lump for now.
+        optionalBlockExtent = Math.max(optionalBlockExtent, when.end());
+      }
+      if (containsMandatoryAttendees) {
+        mandatoryBlockExtent = Math.max(mandatoryBlockExtent, when.end());
+      }
+    }
+
+    // At this point there are no more events, so remaining might be a contiguous potential window.
```

Comment:
(nitpicking) Is it possible to add some tricks here so below code is no longer needed, e.g., add a empty event to the tail of the event list?

Reply:
Thanks for the idea! Managed to modify this.